lsat/interceptor.go

```diff
@@ -0,0 +1,184 @@
+	MaxRoutingFee = 10
+)
+
+var (
+	// authHeaderRegex is the regular expression the payment challenge must
+	// match for us to be able to parse the macaroon and invoice.
+	authHeaderRegex = regexp.MustCompile(
+		"LSAT macaroon='(.*?)' invoice='(.*?)'",
+	)
+)
+
+// Interceptor is a gRPC client interceptor that can handle LSAT authentication
+// challenges with embedded payment requests. It uses a connection to lnd to
+// automatically pay for an authentication token.
+type Interceptor struct {
+	lnd   *lndclient.LndServices
+	store Store
+}
+
+// NewInterceptor creates a new gRPC client interceptor that uses the provided
+// lnd connection to automatically acquire and pay for LSAT tokens, unless the
+// indicated store already contains a usable token.
+func NewInterceptor(lnd *lndclient.LndServices, store Store) (*Interceptor,
+	error) {
+
+	return &Interceptor{
+		lnd:   lnd,
+		store: store,
+	}, nil
+}
+
+// UnaryInterceptor is an interceptor method that can be used directly by gRPC
+// for unary calls.
+// If the store contains a token, it is attached as credentials to every call
+// before patching it through. The response error is also intercepted for every
+// call. If there is an error returned and it is indicating a payment challenge,
+// a token is acquired and paid for automatically. The original request is then
+// repeated back to the server, now with the new token attached.
+func (i *Interceptor) UnaryInterceptor(ctx context.Context, method string,
+	req, reply interface{}, cc *grpc.ClientConn, invoker grpc.UnaryInvoker,
+	opts ...grpc.CallOption) error {
+
+	addLsatCredentials := func(token *Token) error {
+		macaroon, err := token.PaidMacaroon()
+		if err != nil {
+			return err
+		}
+		opts = append(opts, grpc.PerRPCCredentials(
+			macaroons.NewMacaroonCredential(macaroon),
+		))
+		return nil
+	}
+
+	// If we already have a token, let's append it.
+	if i.store.HasToken() {
+		lsat, err := i.store.Token()
+		if err != nil {
+			return err
+		}
+		if err = addLsatCredentials(lsat); err != nil {
+			return err
+		}
+	}
+
+	// We need a way to extract the response headers sent by the
+	// server. This can only be done through the experimental
+	// grpc.Trailer call option.
+	// We execute the request and inspect the error. If it's the
+	// LSAT specific payment required error, we might execute the
+	// same method again later with the paid LSAT token.
+	trailerMetadata := &metadata.MD{}
+	opts = append(opts, grpc.Trailer(trailerMetadata))
+	err := invoker(ctx, method, req, reply, cc, opts...)
+
+	// Only handle the LSAT error message that comes in the form of
+	// a gRPC status error.
+	if isPaymentRequired(err) {
+		lsat, err := i.payLsatToken(trailerMetadata)
+		if err != nil {
+			return err
+		}
+		if err = addLsatCredentials(lsat); err != nil {
+			return err
+		}
+
+		// Execute the same request again, now with the LSAT
+		// token added as an RPC credential.
+		return invoker(ctx, method, req, reply, cc, opts...)
```

Comment:
How does this flow handle expiration? So the case that you have a valid token, but it has expired (as they all may for the initial versions, so we aren't locked in). I guess in the end, it depends on how we communicate expiry from the server-side.

Reply:
Yes, that was my assumption also that we would just return "payment required" in that case too and the client would get a new token.
In a future PR we can add a command line option like `--no-auto-lsat`.